### ISSUE TITLE
ElementArrayFinder.map return type too broad

### ISSUE THREAD

**Issue by RomkeVdMeulen** [angular/protractor#4551]
Bug report
---

- TypeScript Version: `2.5.2`

In the current typing of method ElementArrayFinder.map(), the return type of the mapping function is [marked as `T | any`](https://github.com/angular/protractor/blob/5.2.0/lib/element.ts#L653). I'm not sure why this was ever implemented this way: just using `T` as the return type should accomodate mapping functions with any return type, and even function that don't explicitly return anything. By adding the `any` option, the generic type information is lost, and the result of the entire `map()` call is interpreted as `promise.Promise<{}[]>`, rather than a promise of an array of the mapping's return type.

Using these types, client code has to explicitly mark the return type to have it validated correctly. E.g.:

```ts
function myParagraphs(): promise.Promise<Paragraph[]> {
	return element.all(by.tagName("p")).map(p => new Paragraph(p!));
}
```

Removing the `any` type from the mapping function's return type should correct this problem and be backwards compatible.

**Comment by john-patterson** [comment]
Yeah this has been a major pain point migrating from Protractor 4 to Protractor 5 for me.
